vortex-ffi/src/sink.rs

```diff
@@ -0,0 +1,80 @@
+use std::ptr::null_mut;
+
+use mpsc::Sender;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+use vortex::ArrayRef;
+use vortex::dtype::DType;
+use vortex::error::{VortexExpect, VortexResult, vortex_err};
+use vortex::stream::{ArrayStreamAdapter, ArrayStreamExt};
+
+use crate::array::vx_array;
+use crate::error::{try_or, vx_error};
+use crate::stream::{ArrayStreamInner, vx_array_stream};
+
+#[allow(non_camel_case_types)]
+/// The `sink` object is a writeable array stream, used to go from an external iterator of values
+/// into a `vx_array_stream`.
+pub struct vx_array_sink {
+    sink: Sender<VortexResult<ArrayRef>>,
+}
+
+#[repr(C)]
+#[allow(non_camel_case_types)]
+/// The result of `vx_array_stream_sink_create`.
+pub struct vx_array_stream_sink_create_result {
+    sink: *mut vx_array_sink,
+    stream: *mut vx_array_stream,
+}
+
+/// Opens a writable array stream, where sink is used to push values into the stream.
+/// To close the stream close the sink with `vx_array_sink_close`.
+#[unsafe(no_mangle)]
+pub unsafe extern "C-unwind" fn vx_array_stream_sink_create(
+    dtype: *const DType,
+    error: *mut *mut vx_error,
+) -> vx_array_stream_sink_create_result {
+    try_or(
+        error,
+        vx_array_stream_sink_create_result {
+            sink: null_mut(),
+            stream: null_mut(),
+        },
+        || {
+            let file_dtype = unsafe { dtype.as_ref().vortex_expect("null dtype") };
+
+            let (tx, rx) = mpsc::channel(32);
```

Comment:
I guess the queue needs to be blocking bc we call this from C++. Is there a rationale for going with 32?

Reply:
Nope, just a default